Source/WebCore/page/Location.cpp

```diff
@@ -255,6 +255,11 @@ ExceptionOr<void> Location::replace(LocalDOMWindow& activeWindow, LocalDOMWindow
     if (!activeWindow.document()->canNavigate(dynamicDowncast<LocalFrame>(frame.get()), completedURL))
         return Exception { SecurityError };
 
+    // If the loader for activeWindow's frame (browsing context) has no outgoing referrer, then set the loader's
+    // outgoing referrer to the URL of firstWindow's frame's (browsing context's) Document — which is the
+    // incumbent global object's associated Document. https://html.spec.whatwg.org/#location-object-navigate
+    if (activeWindow.frame()->loader().outgoingReferrer().isEmpty())
+        activeWindow.frame()->loader().setOutgoingReferrer(firstFrame->document()->urlForBindings());
     // We call LocalDOMWindow::setLocation directly here because replace() always operates on the current frame.
     frame->window()->setLocation(activeWindow, completedURL, SetLocationLocking::LockHistoryAndBackForwardList);
```

Comment:
So why don't we do the same thing as `assign()` and pass _firstWindow_ as second argument here? That way `setLocation()` would take care of it, no?

Reply:
> So why don't we do the same thing as `assign()` and pass _firstWindow_ as second argument here? That way `setLocation()` would take care of it, no?

Maybe? I’ll try it and see

Reply:
…and to be clear, along with that _“We call LocalDOMWindow::setLocation directly here because replace() always operates on the current frame”_ says, it seems like the more-specific reason that `Location::replace` can’t do the same thing as what `Location::assign` is doing is: `Location::replace` needs to make `DOMWindow::setLocation` get called with `SetLocationLocking::LockHistoryAndBackForwardList` rather than with the default `SetLocationLocking::LockHistoryBasedOnGestureState` that `Location::assign` and all the other related `Location` member functions end up making `DOMWindow::setLocation` get called with.

Reply:
I'll defer to @cdumez, but from an HTML Standard perspective `assign()` and `replace()` call into the same underlying algorithm in the same way, with `replace()` indeed having a different history value (and no security check).

Reply:
> from an HTML Standard perspective `assign()` and `replace()` call into the same underlying algorithm in the same way

OK, given that — and also given that testing reveals `location.assign` has the same interop problem as `location.replace`: no `Referer` header when a document calls it from an iframe — I’ve updated the patch to move the change out of the `Location::replace` code and into the `LocalDOMWindow::setLocation` code. So that causes the same (code) algorithm to now get applied for both `location.replace` and `location.assign` handling.

I also updated the WPT test to include a `location.assign` case.